### PR TITLE
add s token manager modules

### PR DESCRIPTION
**Pull request by Akira-Taniguchi**
後で

### CODE REVIEW

File: contracts/ISTokensDescriptor.sol [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: MPL-2.0
+pragma solidity 0.8.4;
+
+import {ISTokensManager} from "@devprotocol/i-s-tokens/contracts/interface/ISTokensManager.sol";
+
+interface ISTokensDescriptor {
+	function getTokenURI(ISTokensManager.StakingPosition memory _position)
+		external
```
Comment (aggre) [inline contracts/ISTokensDescriptor.sol:8]: Use `internal`

File: contracts/STokenManager.sol [hunk 2]
```diff
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: MPL-2.0
+pragma solidity 0.8.4;
+
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {Counters} from "@openzeppelin/contracts/utils/Counters.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {ISTokensManager} from "@devprotocol/i-s-tokens/contracts/interface/ISTokensManager.sol";
+import {IAddressConfig} from "./IAddressConfig.sol";
+import {STokensManagerStorage} from "./STokensManagerStorage.sol";
+import {ISTokensDescriptor} from "./ISTokensDescriptor.sol";
```
Comment (aggre) [inline contracts/STokenManager.sol:10]: Use inheritance instead of import

File: contracts/STokenManager.sol [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		require(
+			IAddressConfig(config).lockup() == _msgSender(),
+			"illegal access"
+		);
+		_;
+	}
+
+	function setDescriptor(address _descriptor) external onlyOwner {
+		descriptor = _descriptor;
+	}
```
Comment (aggre) [inline contracts/STokenManager.sol:38]: Remove

File: contracts/STokenManager.sol [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+	function tokenURI(uint256 _tokenId)
+		public
+		view
+		override
+		returns (string memory)
+	{
+		require(_exists(_tokenId), "not found");
+		ISTokensDescriptor sTokensDescriptor = ISTokensDescriptor(descriptor);
+		return sTokensDescriptor.getTokenURI(getStoragePositionsV1(_tokenId));
```
Comment (aggre) [inline contracts/STokenManager.sol:48]: Change the calling way to call inherited methods

File: contracts/STokenManager.sol [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	function mint(MintParams calldata _params)
+		external
+		onlyLockup
+		override
+		returns (uint256 tokenId, StakingPosition memory position)
+	{
+		_tokenIds.increment();
+		uint256 newItemId = _tokenIds.current();
+		_safeMint(_params.owner, newItemId);
+		// TODO 0でいいのか後で確認する
```
Comment (aggre) [inline contracts/STokenManager.sol:60]: No problem :)

File: contracts/STokenManager.sol [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+		onlyLockup
+		override
+		returns (StakingPosition memory position)
+	{
+		StakingPosition memory currentPosition = getStoragePositionsV1(
+			_params.tokenId
+		);
+		currentPosition.amount = _params.amount;
+		currentPosition.price = _params.price;
+		currentPosition.historical = _params.historical;
```
Comment (aggre) [inline contracts/STokenManager.sol:83]: Modify the StakingPosition interface for the latest DIP

File: contracts/STokensDescriptor.sol [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import {Base64} from "@devprotocol/util-contracts/contracts/utils/Base64.sol";
+import {ISTokensDescriptor} from "./ISTokensDescriptor.sol";
+
+contract STokensDescriptor is ISTokensDescriptor {
+	using Base64 for bytes;
+	using AddressLib for address;
+	using Strings for uint256;
+
+	function getTokenURI(ISTokensManager.StakingPosition memory _position)
+		external override
```
Comment (aggre) [inline contracts/STokensDescriptor.sol:16]: `external` to `internal`

File: contracts/STokensManagerStorage.sol [hunk 8]
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: MPL-2.0
+pragma solidity 0.8.4;
+
+import {ISTokensManager} from "@devprotocol/i-s-tokens/contracts/interface/ISTokensManager.sol";
+
+contract STokensManagerStorage {
```
Comment (aggre) [inline contracts/STokensManagerStorage.sol:6]: Storage is stored in the Proxy contract, so why need this contract?

**Review by aggre — COMMENTED** [review]
I've reviewed some points

File: contracts/ISTokensDescriptor.sol (hunk 1 above)
Comment (Akira-Taniguchi) [reply contracts/ISTokensDescriptor.sol:8]: The size of the contract is so large that it cannot be deployed.
So we decided to separate the STokensDescriptor.
Is this a problem?
@aggre

File: contracts/STokenManager.sol (hunk 3 above)
Comment (Akira-Taniguchi) [reply contracts/STokenManager.sol:38]: https://github.com/dev-protocol/s-tokens/pull/1#discussion_r693687910

Comment (Akira-Taniguchi) [reply contracts/STokenManager.sol:38]: https://github.com/dev-protocol/s-tokens/pull/1#discussion_r693687910

For the same reason.

File: contracts/STokenManager.sol (hunk 4 above)
Comment (Akira-Taniguchi) [reply contracts/STokenManager.sol:48]: #1 (comment)

For the same reason.

@aggre

File: contracts/STokenManager.sol (hunk 5 above)
Comment (Akira-Taniguchi) [reply contracts/STokenManager.sol:60]: thanks @aggre

File: contracts/STokensManagerStorage.sol (hunk 8 above)
Comment (Akira-Taniguchi) [reply contracts/STokensManagerStorage.sol:6]: #1 (comment)

For the same reason.

File: contracts/STokensDescriptor.sol (hunk 7 above)
Comment (Akira-Taniguchi) [reply contracts/STokensDescriptor.sol:16]: https://github.com/dev-protocol/s-tokens/pull/1#discussion_r693687910

For the same reason.

@aggre

File: contracts/STokensManagerStorage.sol (hunk 8 above)
Comment (Akira-Taniguchi) [reply contracts/STokensManagerStorage.sol:6]: I will delete it
@aggre

File: contracts/STokenManager.sol (hunk 2 above)
Comment (Akira-Taniguchi) [reply contracts/STokenManager.sol:10]: https://github.com/dev-protocol/s-tokens/pull/1#discussion_r693687910

For the same reason.
@aggre

**Comment by aggre** [comment]
[sTokens](https://trello.com/c/H6bNUScR/1-stokens)

**Pull request merged**